### PR TITLE
improve(pricefeed): Add getHistoricalPricePeriods to BasketSpread and Uniswap

### PR DESCRIPTION
**Pull request by nicholaspai**
Signed-off-by: Nick Pai <npai.nyc@gmail.com>

**Motivation**

Part of #2461

**Summary**

Implements `getHistoricalPricePeriods() => [time, price]` for Uniswap and BasketSpread price feeds.

**Testing**

TODO: Does this work for `affiliates`?

- [x]  Ran end-to-end test, running the code as in production
- [ ]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [x]  Untested

### CODE REVIEW

File: packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js [hunk 1]
```diff
@@ -143,6 +143,27 @@ class BasketSpreadPriceFeed extends PriceFeedInterface {
...
   }
 
+  // Note: This method will arbitrarily fail if the first baseline pricefeed has not implemented `getHistoricalPricePeriods`
+  getHistoricalPricePeriods() {
+    // Arbitrarily use the first baseline pricefeed's price periods as the index for price periods to return.
+    // This function is hacky and makes an assumption that the baseline pricefeed is a Medianizer and therefore
+    // returns price periods as [time, price].
+    const pricePeriods = this.baselinePriceFeeds[0].getHistoricalPricePeriods();
+
+    return pricePeriods.map(_pricePeriod => {
```
Comment (nicholaspai) [inline packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:153]: Is this sufficient for dev mining calcs?

File: packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js [hunk 2]
```diff
@@ -143,6 +143,27 @@ class BasketSpreadPriceFeed extends PriceFeedInterface {
...
+      let timestamp = _pricePeriod[0];
+      // If we can fetch historical price for this timestamp we'll add it to the price periods array.
+      this.getHistoricalPrice(timestamp)
```
Comment (daywiss) [inline packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:156]: i dont think this is gonna work right. the map is going to return an array of promises

File: packages/financial-templates-lib/src/price-feed/UniswapPriceFeed.js [hunk 3]
```diff
@@ -76,6 +76,15 @@ class UniswapPriceFeed extends PriceFeedInterface {
     }
   }
 
+  getHistoricalPricePeriods() {
```
Comment (daywiss) [inline packages/financial-templates-lib/src/price-feed/UniswapPriceFeed.js:79]: this looks good to me, but may need to worry about how this works with the medianizer

File: packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js (hunk 1 above)
Comment (daywiss) [reply packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:153]: i think the intention is right yea. but test the implementation

Comment (nicholaspai) [reply packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:153]: What's the best way to test this?

File: packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js (hunk 2 above)
Comment (nicholaspai) [reply packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:156]: Doesn't the .then() make it synchrous?

**Review by chrismaree — COMMENTED** [review]
looks correct but I think @daywiss will need to double check it works as expected with the affiliates package.

File: packages/core/scripts/identifier-umip/1_Propose.js [hunk 4]
```diff
@@ -45,7 +45,8 @@ async function runExport() {
     };
   });
 
-  await governor.propose(transactions, { from: proposerWallet });
+  const proposal = await governor.propose(transactions, { from: proposerWallet, gas: 2000000, gasPrice: 300000000000});
+  console.log(await web3.eth.getTransaction(proposal.receipt.transactionHash))
```
Comment (nicholaspai) [inline packages/core/scripts/identifier-umip/1_Propose.js:49]: Leftover work from another PR?

File: yarn.lock [hunk 5]
```diff
@@ -3830,31 +3830,6 @@
   resolved "https://registry.yarnpkg.com/@protobufjs/utf8/-/utf8-1.1.0.tgz#a777360b5b39a1a2e5106f8e858f2fd2d060c570"
   integrity sha1-p3c2C1s5oaLlEG+OhY8v0tBgxXA=
 
-"@sendgrid/client@^6.5.5":
```
Comment (nicholaspai) [inline yarn.lock:3833]: Is this yarn.lock meant to be included?

**Comment by nicholaspai** [comment]
@daywiss is this ready to move on from draft mode?

File: packages/core/scripts/identifier-umip/1_Propose.js (hunk 4 above)
Comment (daywiss) [reply packages/core/scripts/identifier-umip/1_Propose.js:49]: yea thats my bad

**Comment by daywiss** [comment]
@nicholaspai let me do another once over and ill upgrade the pr

File: yarn.lock (hunk 5 above)
Comment (daywiss) [reply yarn.lock]: dont know why the lock file changed, but ill remove it

File: packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js [hunk 6]
```diff
@@ -134,7 +135,55 @@ class BasketSpreadPriceFeed extends PriceFeedInterface {
...
+      assert(result, "no closest time found");
+      return this.toBN(result[1]);
+    };
+  }
+  // This function does something similar to get historicalprice, but does not have the luxury of only caring about a
+  // single point in time. It has to run the basketspread price across all timestamps available. This is complicated
+  // as there are multiple price histories which we must search through at each matching timestamp to find the closets
+  // prices to add into the basket calculation.
+  // Returns data in the form of [[time,price]]
+  getHistoricalPricePeriods() {
```
Comment (daywiss) [inline packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:161]: this function was refactored and its quite complicated. the idea here is that there are multiple historical price feeds. its possible these feeds all have different timestamps as well. we need to somehow go through the time, pull the nearest price from each feed, then pass it into the `getSpreadFromBasketPrices` function. 

In order to do this, ive added this `closestTime` function as a  helper to search for closest timestamp within each price feed.

File: packages/financial-templates-lib/src/price-feed/CryptoWatchPriceFeed.js [hunk 7]
```diff
@@ -136,17 +136,15 @@ class CryptoWatchPriceFeed extends PriceFeedInterface {
...
+        return [historicalPrice.closeTime, historicalPrice.closePrice];
+      });
     else
       return this.historicalPricePeriods.map(historicalPrice => {
-        return {
-          ...historicalPrice,
-          openPrice: this._invertPriceSafely(historicalPrice.openPrice),
-          closePrice: this._invertPriceSafely(historicalPrice.closePrice)
-        };
+        return [historicalPrice.closeTime, this._invertPriceSafely(historicalPrice.closePrice)];
```
Comment (daywiss) [inline packages/financial-templates-lib/src/price-feed/CryptoWatchPriceFeed.js:145]: refactored to be similar to medianizer output [[time,price]]

File: yarn.lock (hunk 5 above)
Comment (nicholaspai) [reply yarn.lock]: You should merge master I merged a new `yarn.lock`

File: packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js [hunk 8]
```diff
@@ -134,7 +135,55 @@ class BasketSpreadPriceFeed extends PriceFeedInterface {
       return this._getSpreadFromBasketPrices(experimentalPrices, baselinePrices, denominatorPrice);
     }
   }
+  // This searches for closest time in a list of [time,price] data. Based on code in affiliates models/prices.
+  closestTime(list) {
```
Comment (nicholaspai) [inline packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:139]: You might want to add a comment explaining the expected shape of `list`

Comment (daywiss) [reply packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:139]: there is a comment above but i think its not technically correct. i will update it to make it more clear

**Comment by daywiss** [comment]
need to get tests to pass, but successfully ran dev mining based on these changes. should be able to go in soon

**Review by daywiss — APPROVED** [review]
approving since I ran an e2e style test which touched on uni, medianizer, basket spread, cryptowatch and domfi price feeds and seemed to work fine.

**Pull request merged**